### ISSUE TITLE
Definitionally Equal Inductive/Record outputs

### ISSUE THREAD

**Issue by soundlogic2236** [coq/coq#18456]
#### Description of the problem

With primitive projections, we can create types that very transparently wrap an underlying type, but create new notations that cannot at least easily be duplicated without them. However, we must choose between getting these notational forms, using definitional aliases, and using notations.

Consider the following:
```coq
Record prod@{*a *b|} (A : Type@{a}) (B : Type@{b}) : Type@{max(a, b)} := pair {
  fst : A;
  snd : B;
}.
```
Here we have a sort polymorphic product, a handy thing added recently. 
Quite wonderfully, it covers the `Prop` product (aka `and`), and the `SProp` product.

However, if we want to use this we then have some choices to make.

We could do things via notational aliases:
```coq
Notation and := prod@{Prop|Set Set}
Notation conj := pair@{Prop|Set Set}
```
Alas, these have multiple issues. First off, they won't print properly, printing as the underlying terms. Secondly, they can't be registered:
```coq
Fail Register and as core.and.type.
(* "and is bound to a notation that does not denote a reference." *)
```
All right, we could just use definitions. After all, we have first class types!
This will print fine... as long as you don't unfold it. It can be registered (though I don't know how well the actual code will handle things like this). It is what is currently done with `iff` and `and` in the standard library, so it has... history. 
We can make them opaque, but that then results in things like `proj1 (conj x y)` failing to reduce to `x`.

We could create a completely new record:
```coq
Record and@{|} (A B : Prop) : Prop := conj {
  proj1 : A;
  proj2 : B;
}.
```
This will no longer `cbv` into the other, though now we have to redefine everything. Coercions can help with that, but... it is still a lot of duplication. Further, if our type isn't itself something based on primitive projections, we cannot write the conversions we really want, as everything will have eta expansions and sometimes even fixpoints around. 

We could create a wrapper record:
```coq
Record and@{|} (A B : Prop) : Prop := prod_to_and {
  and_to_prod : prod A B;
}.
```
Now our conversions won't have eta expansions and fixpoints, and we won't reduce via cbv... in the type. If we define our `conj` it will still `cbv` into `prod_to_and` combined with `pair`. 

But really... why *must* `and` be either reducible to `prod` *or* not convertible with it? 

Plenty of models validate propositional equality between (correctly monosorted) `prod` and `and`. All univalent ones, but cardinality models and simple ones where (primitive projection) `Record` is interpreted as `Definition` with an underlying structurally typed record system.

Making all types of the same 'shape' unifiable would be radical, and also rather delicate (would switching the order of `cons` and `nil` break your code? Would the output of converting a bool to another type change if `True` and `False` were switched?).

However, making the `Inductive` and `Record` commands _able_ to output something unifiable with another type would, while potentially a lot of implementation work, I believe be a completely conservative extension to the theory of coq. Everything here can be done with a pile of definitions, it "just" comes out as changing reduction behavior while leaving conversion exactly the same, and reduction behavior (as long as it only reduces to another convertable thing) is irrelevant. Already, we have `fun x => f x` and `f` convertible but neither reduces to the other. `Opaque` definitions are convertible but don't reduce.

Implementation and syntax on the other hand, while not very relevant to the underlying theory, would potentially be a much bigger ask. In the limit, one could imagine this subsuming already major additions like custom pattern matching.

Even if it was limited to straightforwardly simple shape + specialization, that still would leave implementation questions like how to alter the unification algorithm, and how to specify which constructors go to which constructors, which projections to go which projections, and so on.

Alternatively, some way of controlling opaque definitions so that they can reduce in matches or projections might work. Since this is purely a question of reduction behavior *sufficient* control over reductions should be equivalent with at worst the cost of extra boilerplate. It would be undesirable though if it was impossible or near impossible to write new functions on `fooList`s without interacting with them the fact that `fooList` is defined as `list Foo`, such as if one had to write extra commands to inform the system that your new function should reduce on the "constructors" of `fooList` rather than it generally being able to understand that it is a function that matches on a `fooList` and `fooNil` is a "constructor" of `fooList`. With a sufficient system here, such an `Inductive` declaration would be pure syntactic sugar for the setup.

In conclusion: not having to choose between a pile of finicky definitions for one's specialized `list Foo` and losing the possibility of a pair of functions `fooList -> list Foo` and `list Foo -> fooList` which when composed give the identity function would I think be nice.

**Comment by SkySkimmer** [comment]
See also https://github.com/coq/coq/issues/5293 I guess
